### PR TITLE
ENH Allow creating temporary templates

### PR DESCRIPTION
**Pull request by GuySartorelli**
Required for silverstripe/silverstripe-elemental#980

Provides a way to safely add a temporary template for a given scenario, which gets cleaned up afterwards (consistent with the way temporary config is added)

### CODE REVIEW

File: src/Context/FixtureContext.php [hunk 1]
```diff
@@ -966,11 +1008,29 @@ protected function clearConfigFiles()
         $this->activatedConfigFiles = [];
     }
 
+    protected function clearTemplateFiles(): void
+    {
+        // No files to cleanup
+        if (empty($this->createdTemplatePaths)) {
+            return;
+        }
+
```
Comment (emteknetnz) [inline src/Context/FixtureContext.php:1017]:
```suggestion
```
I'm not a fan of extra lines line this, it's still easy to read without them and it just increases the need for scrolling

File: src/Context/FixtureContext.php [hunk 2]
```diff
@@ -687,6 +695,42 @@ class_exists($extension) && is_subclass_of($extension, Extension::class),
         }
     }
 
+    /**
+     * Get the module that represents the application.
+     */
+    protected function getAppModule(): Module
+    {
+        $project = ModuleManifest::config()->get('project') ?: 'mysite';
```
Comment (emteknetnz) [inline src/Context/FixtureContext.php:703]:
```suggestion
        $project = ModuleManifest::config()->get('project') ?: 'app';
```
More likely to be `app` these days?

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I'd question why we even need to have mock templates when we're testing the "out of the box" experience

Wouldn't a better solution be to just add an actual template to admin or similar module if there's a missing template?

File: src/Context/FixtureContext.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Context/FixtureContext.php:703]: This is just copied from `getDestinationConfigFolder()`, which I've refactored it out from. I can change it if you want, but that's the context.

File: src/Context/FixtureContext.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Context/FixtureContext.php:1017]: Fair enough. I was copying the style of `clearConfigFiles()` above. Shall I remove the extra line from both? It'd be nice if we had some policy (extended code style rules or something) that tell us when new lines are or are not appropriate - I reviewed a PR with a similar new line added and questioned it and one of the core committers said they prefer them in there so 🤷‍♀️

**Comment by GuySartorelli** [comment]
> I'd question why we even need to have mock templates when we're testing the "out of the box" experience
> Wouldn't a better solution be to just add an actual template to admin or similar module if there's a missing template?

The thing is we're testing the preview of the front-end here - and really the theme used to include content on the front-end is very much not out of the box. It's not that a template is "missing" per se - it's just that we don't control the front-end to the extend we control the CMS. We do have a couple of themes but the tests shouldn't be dependent on a specific theme being included. So this allows tests to add front-end templates for testing front-end stuff (like, in the case of the PR that spawned this one, elemental blocks).

I can try to do something with frameworktest if you'd prefer - but elemental doesn't currently have frameworktest as a dependency, plus this allows for the temporary addition of what is effectively a test fixture without needing to include additional libraries, so I kinda like this approach - plus it's consistent with the way config is already being added.

File: src/Context/FixtureContext.php [hunk 3]
```diff
@@ -687,6 +695,42 @@ class_exists($extension) && is_subclass_of($extension, Extension::class),
...
+    }
+
+    /**
+     * Get the main folder for the application.
+     */
+    protected function getDestinationTemplatePath(string $filename, bool $createMissingDirs): string
+    {
+        $app = $this->getAppModule();
+        $basePath = $app->getPath();
+        $destPath = $app->getResource("templates/{$filename}")->getPath();
```
Comment (GuySartorelli) [inline src/Context/FixtureContext.php:716]: There's an argument to be made that this should be creating a new temporary test theme instead - which I'd be open to doing if that's preferred.

**Comment by GuySartorelli** [comment]
I thought of a much simpler way of handling this for the one test that needed it, so this PR isn't needed now.

**Pull request closed without merging**